### PR TITLE
feat: pkg/util/helm Enhancement (sub task of pkg/util Enhancement #83…

### PR DESCRIPTION
**Pull request by jxs1211**
## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description
This is an enhancement for pkg/util/helm, which include:
- refactor the 'NewHelm' for supporting customized Client and easier mocking
- add ut
- some other refactors

## Related Issues
see https://github.com/devstream-io/devstream/issues/838 and https://github.com/devstream-io/devstream/issues/833 for the details

## New Behavior (screenshots if needed)
ut result:

[going@dev devstream]$ go tool cover -func=coverage.out
github.com/devstream-io/devstream/pkg/util/helm/helm.go:22:             NewHelm                         82.4%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:95:             WithEntry                       100.0%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:101:            WithChartSpec                   100.0%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:107:            WithClient                      100.0%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:113:            AddOrUpdateChartRepo            100.0%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:117:            InstallOrUpgradeChart           100.0%
github.com/devstream-io/devstream/pkg/util/helm/helm.go:122:            UninstallHelmChartRelease       100.0%
github.com/devstream-io/devstream/pkg/util/helm/state.go:15:            ToStringInterfaceMap            80.0%
github.com/devstream-io/devstream/pkg/util/helm/state.go:38:            AddDeployment                   100.0%
github.com/devstream-io/devstream/pkg/util/helm/state.go:48:            AddDaemonset                    100.0%
github.com/devstream-io/devstream/pkg/util/helm/state.go:58:            AddStatefulset                  100.0%
github.com/devstream-io/devstream/pkg/util/helm/validation.go:6:        defaults                        100.0%
github.com/devstream-io/devstream/pkg/util/helm/validation.go:13:       validate                        100.0%
github.com/devstream-io/devstream/pkg/util/helm/validation.go:17:       DefaultsAndValidate             100.0%
total:                                                                  (statements)                    91.1%

### CODE REVIEW

**Comment by hxcGit** [comment]
@jxs1211  Please fix the build error.

**Comment by hxcGit** [comment]
@jxs1211  hi, here is the reason.
image
You use `mockgen` in your code which hasn't been installed in the workflow machine.

**Comment by aFlyBird0** [comment]
Here is the error: @jxs1211 
image

**Comment by jxs1211** [comment]
@hxcGit the building passed this time!

File: pkg/util/helm/helm.go [hunk 1]
```diff
@@ -80,34 +76,52 @@ func NewHelm(param *HelmParam) (*Helm, error) {
...
-
-	helm := &Helm{
+	h := &Helm{
 		Entry:     entry,
-		ChartSpec: chartSpec,
+		ChartSpec: spec,
 		Client:    hClient,
 	}
 
-	if err = helm.AddOrUpdateChartRepo(*helm.Entry); err != nil {
```
Comment (hxcGit) [inline pkg/util/helm/helm.go:90]: You separate `AddOrUpdateChartRepo()` out to `func (h *Helm) AddOrUpdateChartRepo(entry repo.Entry) error`. But without applying it.

File: pkg/util/helm/interface.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package helm
+
+import (
+	helmclient "github.com/mittwald/go-helm-client"
+)
+
+type Client interface {
```
Comment (hxcGit) [inline pkg/util/helm/interface.go:7]: What is the purpose of this interface? Adding some comments will be great. By the way, I can't find where this interface is used.

File: pkg/util/helm/helm_test.go [hunk 3]
```diff
@@ -0,0 +1,388 @@
+package helm
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	mockhelmclient "github.com/devstream-io/devstream/pkg/util/helm/mock"
```
Comment (hxcGit) [inline pkg/util/helm/helm_test.go:10]: Pls make sure the imported libraries are correctly ordered. Like below,
```golang
import (
	"context"
	"errors"
	"reflect"
	"testing"
	"time"

	"github.com/golang/mock/gomock"
	helmclient "github.com/mittwald/go-helm-client"
	"helm.sh/helm/v3/pkg/release"
	"helm.sh/helm/v3/pkg/repo"

	mockhelmclient "github.com/devstream-io/devstream/pkg/util/helm/mock"
)
```
You can use `make fmt` to do this. (gofmt、goimports)

File: pkg/util/helm/helm_test.go [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+	// default
+	got, err := NewHelm(helmParam)
+	if err != nil {
+		t.Errorf("error: %v\n", err)
+	}
+
+	want := &Helm{
+		Entry:     entry,
+		ChartSpec: spec,
+		Client:    got.Client,
```
Comment (hxcGit) [inline pkg/util/helm/helm_test.go:107]: Use `got.Client` to construct `want` may not make sense. This probably leave out the step of `AddOrUpdateChartRepo()`.

**Comment by hxcGit** [comment]
@jxs1211  Many thanks to your contribution. What a great PR!

File: pkg/util/helm/interface.go (hunk 2 above)
Comment (jxs1211) [reply pkg/util/helm/interface.go:7]: @hxcGit This interface is used to generate mockclient by mockgen

File: pkg/util/helm/helm.go (hunk 1 above)
Comment (jxs1211) [reply pkg/util/helm/helm.go:90]: fixed it

File: pkg/util/helm/helm_test.go (hunk 4 above)
Comment (jxs1211) [reply pkg/util/helm/helm_test.go:107]: fixed

**Comment by hxcGit** [comment]
@jxs1211  That's noraml. The reason for build failure is that the test cases in `pkg/util/helm` did not pass. Show as below
image




image

**Comment by hxcGit** [comment]
@steinliber @aFlyBird0  PTAL.

**Review by steinliber — APPROVED** [review]
I think this is very great! Thank you

**Review by hxcGit — APPROVED** [review]

**Comment by hxcGit** [comment]
@jxs1211  Great job!

**Pull request merged**